src/key_schedule.rs

```diff
@@ -208,93 +162,94 @@ impl KeySchedule {
+{
+    hkdf_expand_info(secret, key_type, label, context, |okm| okm.into())
 }
 
-fn _hkdf_expand_label(output: &mut [u8],
-                      secret: &hmac::Key,
-                      label: &[u8],
-                      context: &[u8]) {
-    let label_prefix = b"tls13 ";
-
-    let mut hkdflabel = Vec::new();
-    (output.len() as u16).encode(&mut hkdflabel);
-    ((label.len() + label_prefix.len()) as u8).encode(&mut hkdflabel);
-    hkdflabel.extend_from_slice(label_prefix);
-    hkdflabel.extend_from_slice(label);
-    (context.len() as u8).encode(&mut hkdflabel);
-    hkdflabel.extend_from_slice(context);
-
-    _hkdf_expand(secret, &hkdflabel, output)
+fn hkdf_expand_info<F, T, L>(secret: &hkdf::Prk, key_type: L, label: &[u8], context: &[u8], f: F)
+        -> T
+    where
+        F: for<'b> FnOnce(hkdf::Okm<'b, L>) -> T,
+        L: hkdf::KeyType
+{
+    const LABEL_PREFIX: &[u8] = b"tls13 ";
+
+    let output_len = u16::to_be_bytes(key_type.len() as u16);
+    let label_len = u8::to_be_bytes((LABEL_PREFIX.len() + label.len()) as u8);
+    let context_len = u8::to_be_bytes(context.len() as u8);
+
+    let info = &[&output_len[..], &label_len[..], LABEL_PREFIX, label, &context_len[..], context];
+    let okm = secret.expand(info, key_type).unwrap();
+
+    f(okm)
 }
 
-pub fn derive_traffic_key(hash: &'static digest::Algorithm, secret: &[u8], len: usize) -> Vec<u8> {
-    let hmac_alg = convert_digest_to_hmac_alg(hash);
-    _hkdf_expand_label_vec(&hmac::Key::new(hmac_alg, secret), b"key", &[], len)
+pub(crate) struct PayloadU8Len(pub(crate) usize);
+impl hkdf::KeyType for PayloadU8Len {
+    fn len(&self) -> usize { self.0 }
 }
 
-pub(crate) fn derive_traffic_iv(hash: &'static digest::Algorithm, secret: &[u8]) -> Iv {
-    let hmac_alg = convert_digest_to_hmac_alg(hash);
-    let iv = _hkdf_expand_label_vec(&hmac::Key::new(hmac_alg, secret), b"iv", &[],
-    ring::aead::NONCE_LEN);
-    Iv::new(iv[..].try_into().unwrap())
+impl From<hkdf::Okm<'_, PayloadU8Len>> for PayloadU8 {
+    fn from(okm: hkdf::Okm<PayloadU8Len>) -> Self {
+        let mut r = vec![0u8;okm.len().0];
+        okm.fill(&mut r[..]).unwrap();
+        PayloadU8::new(r)
+    }
+}
+
+pub fn derive_traffic_key(secret: &hkdf::Prk, aead_algorithm: &'static aead::Algorithm)
+    -> aead::UnboundKey
+{
+    hkdf_expand(secret, aead_algorithm, b"key", &[])
+}
+
+pub(crate) fn derive_traffic_iv(secret: &hkdf::Prk) -> Iv {
+    hkdf_expand(secret, IvLen, b"iv", &[])
 }
 
 #[cfg(test)]
 mod test {
     use super::{KeySchedule, SecretKind, derive_traffic_key, derive_traffic_iv};
-    use ring::digest;
+    use ring::{aead, hkdf};
 
     #[test]
     fn smoke_test() {
         let fake_handshake_hash = [0u8; 32];
 
-        let mut ks = KeySchedule::new(&digest::SHA256);
-        ks.input_empty(); // no PSK
-        ks.derive(SecretKind::ResumptionPSKBinderKey, &fake_handshake_hash);
+        let mut ks = KeySchedule::new_with_empty_secret(hkdf::HKDF_SHA256);
```

Comment:
I'm not sure this `smoke_test()` test is even useful anymore, especially considering the existence of the other test. In particular, it's testing `derive_bytes` only but `derive_bytes` is the worst function to use.